pylon-core/src/main/kotlin/io/github/pylonmc/pylon/core/entity/EntityStorage.kt

```diff
@@ -0,0 +1,162 @@
+package io.github.pylonmc.pylon.core.entity
+
+import com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent
+import io.github.pylonmc.pylon.core.addon.PylonAddon
+import io.github.pylonmc.pylon.core.event.PylonEntityDeathEvent
+import io.github.pylonmc.pylon.core.event.PylonEntityLoadEvent
+import io.github.pylonmc.pylon.core.event.PylonEntityUnloadEvent
+import io.github.pylonmc.pylon.core.pluginInstance
+import io.github.pylonmc.pylon.core.registry.PylonRegistry
+import io.github.pylonmc.pylon.core.util.isFromAddon
+import org.bukkit.Bukkit
+import org.bukkit.NamespacedKey
+import org.bukkit.entity.Entity
+import org.bukkit.event.EventHandler
+import org.bukkit.event.Listener
+import org.bukkit.event.entity.EntityDeathEvent
+import org.bukkit.event.world.EntitiesLoadEvent
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+
+object EntityStorage : Listener {
+
+    private const val AUTOSAVE_INTERVAL_TICKS = 60 * 20L
+
+    private val entities: MutableMap<UUID, PylonEntity<*, *>> = ConcurrentHashMap()
+    private val entitiesById: MutableMap<NamespacedKey, MutableSet<PylonEntity<*, *>>> = ConcurrentHashMap()
+
+    val loadedEntities: Collection<PylonEntity<*, *>>
+        get() = entities.values
+
+    // Access to entities, entitiesById fields must be synchronized to prevent them
+    // briefly going out of sync
+    private val entityLock = ReentrantReadWriteLock()
+
+    // TODO implement this properly and actually run it
+    internal fun startAutosaveTask() {
+        Bukkit.getScheduler().runTaskTimer(pluginInstance, Runnable {
+            // TODO this saves all entities at once, potentially leading to large pauses
+            for (entity in entities.values) {
+                entity.write()
+            }
+        }, AUTOSAVE_INTERVAL_TICKS, AUTOSAVE_INTERVAL_TICKS)
+    }
+
+    @JvmStatic
+    fun get(uuid: UUID): PylonEntity<*, *>?
+        = lockEntityRead { entities[uuid] }
+
+    @JvmStatic
+    fun get(entity: Entity): PylonEntity<*, *>?
+        = get(entity.uniqueId)
+
+    @JvmStatic
+    fun <T> getAs(clazz: Class<T>, uuid: UUID): T? {
+        val entity = get(uuid) ?: return null
+        return clazz.cast(entity)
+    }
+
+    @JvmStatic
+    fun <T> getAs(clazz: Class<T>, entity: Entity): T?
+        = getAs(clazz, entity.uniqueId)
+
+    inline fun <reified T> getAs(uuid: UUID): T?
+        = getAs(T::class.java, uuid)
+
+    inline fun <reified T> getAs(entity: Entity): T?
+        = getAs(T::class.java, entity)
+
+    fun getById(id: NamespacedKey): Collection<PylonEntity<*, *>> =
+        if (PylonRegistry.ENTITIES.contains(id)) {
```

Comment:
```suggestion
        if (PylonRegistry.ENTITIES.contains(key)) {
```
corresponding to ^ change